### ISSUE TITLE
agent vs NeonVM state is inconsistent when NeonVM fails

### ISSUE THREAD

**Issue by sharnoff** [neondatabase/autoscaling#23]
During the all-hands demo 2023-01-20, the state of the VM as reported by NeonVM didn't match what the autoscaler-agent (and scheduler) thought it was. To be specific:

* Under high load, the autoscaler-agent decided to scale the VM up from 16 CPU + 64 GB to its maximum of 19 CPU + 76 GB, which was approved by the scheduler.
* Sometime after that, I looked at the state from NeonVM, and it didn't match — NeonVM reported 18 CPU + 66 GB, with the autoscaler-agent remaining constant.
* The autoscaler-agent's runner for the VM did not emit any logs relating to the request to NeonVM (so: presumably it did not error, unless the request handling silently ignored a failure).

Didn't have a chance to check from inside the VM (so: I don't know which is correct). Haven't checked whether this is easy to replicate.

---

Extra details:
* VM name: `compute-node-ep-throbbing-wave`
* Time of attempted scaling 16 -> 19 CU: `2023-01-20 17:09:32`

A helpful command to filter downloaded logs for the autoscaler-agent's runner for the VM:

```sh
grep '^{' log.txt \
    | jq -r '.log' \
    | grep -E '^(I|W|E|F).*Runner default:compute-node-ep-throbbing-wave-612027-7zz9t' \
    | sort > log
```

**Comment by sharnoff** [comment]
Looks like NeonVM got stuck trying & failing to reconcile the state. Found some errors, all along the lines of `"a used vhost backend has no free memory slots left"`. Here's the logs from the attempted increase from 16 CPU + 64 GB to 19 CPU + 76 GB:

<details><summary>Expand for: <code>neonvm-controller</code> logs from <code>17:09:32</code> to <code>17:10:30</code></summary>
<pre>
1.6742345720475886e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/mutate-vm-neon-tech-v1-virtualmachine", "UID": "c97a6ec3-4cd8-442b-9ac4-a4a533b44561", "kind": "vm.neon.tech/v1, Kind=VirtualMachine", "resource": {"group":"vm.neon.tech","version":"v1","resource":"virtualmachines"}}
1.6742345720478792e+09	INFO	virtualmachine-resource	default	{"name": "compute-node-ep-throbbing-wave-612027"}
1.6742345720483289e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/mutate-vm-neon-tech-v1-virtualmachine", "code": 200, "reason": "", "UID": "c97a6ec3-4cd8-442b-9ac4-a4a533b44561", "allowed": true}
1.6742345720530186e+09	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/validate-vm-neon-tech-v1-virtualmachine", "UID": "7372a110-d3ba-47ba-9b60-5202b5c8706e", "kind": "vm.neon.tech/v1, Kind=VirtualMachine", "resource": {"group":"vm.neon.tech","version":"v1","resource":"virtualmachines"}}
1.6742345720537157e+09	INFO	virtualmachine-resource	validate update	{"name": "compute-node-ep-throbbing-wave-612027"}
1.6742345720537498e+09	DEBUG	controller-runtime.webhook.webhooks	wrote response	{"webhook": "/validate-vm-neon-tech-v1-virtualmachine", "code": 200, "reason": "", "UID": "7372a110-d3ba-47ba-9b60-5202b5c8706e", "allowed": true}
1.6742345721386282e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 17 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683136"}, "reason": "CpuInfo"}
1.6742345722014503e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 65Gi memory	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683136"}, "reason": "MemoryInfo"}
1.674234572252486e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 18 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683140"}, "reason": "CpuInfo"}
1.6742345723150055e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 66Gi memory	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683140"}, "reason": "MemoryInfo"}
1.6742345723658578e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345724576654e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "3e464d44-f2fc-4476-966f-278516e90884", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345724577212e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345724652631e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345725088322e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345725088089e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "01bcbe54-c486-4a5c-bb13-1d1d9f453750", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345725217192e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345725692983e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "52bdbe5c-8102-445b-b360-336f740020e4", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345725693376e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.674234572591805e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345727223706e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "48a3fdda-6e86-4d8b-8935-983754fd583a", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345727224314e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.674234572765575e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345728241513e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "db3677ea-dc87-4360-ab41-ff39629a95c3", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345728241978e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.674234572907013e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345729654732e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "93e472d2-a327-4315-a7e3-b0087c3d5d98", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.674234572965548e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345731290314e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345731885724e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "427b8288-ed67-42af-9d25-c4143f182735", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345731886346e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345735116875e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.674234573570796e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "e6c49f82-c79b-4b59-b2d4-8c61792485f4", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345735708559e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345742143285e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345742734363e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "9c977472-7526-4bdc-9d31-b6b8cf273dfc", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345742734973e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345755565357e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345756161203e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "737760c9-7de4-46bd-99d2-71d9f25a9937", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345756161735e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345781789484e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345782380307e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "085bcb7a-0ce4-4472-be07-a74466687868", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345782380426e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345833613765e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345834209912e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "a956e58f-a3e6-4af3-9d8e-6f440bd4a579", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742345834210522e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742345936641603e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742345937231417e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "344148ab-bb75-4306-b7c9-b1559501b291", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.67423459372323e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
1.6742346142064593e+09	DEBUG	events	VirtualMachine compute-node-ep-throbbing-wave-612027 uses 19 cpu cores	{"type": "Normal", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "CpuInfo"}
1.6742346142654989e+09	ERROR	Reconciler error	{"controller": "virtualmachine", "controllerGroup": "vm.neon.tech", "controllerKind": "VirtualMachine", "VirtualMachine": {"name":"compute-node-ep-throbbing-wave-612027","namespace":"default"}, "namespace": "default", "name": "compute-node-ep-throbbing-wave-612027", "reconcileID": "a0a7da2c-f91f-4e05-bf10-90b3c0684466", "error": "a used vhost backend has no free memory slots left"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
1.6742346142655585e+09	DEBUG	events	Failed to reconcile (compute-node-ep-throbbing-wave-612027): a used vhost backend has no free memory slots left	{"type": "Warning", "object": {"kind":"VirtualMachine","namespace":"default","name":"compute-node-ep-throbbing-wave-612027","uid":"48fc066a-c35e-487e-8f77-4169ac539006","apiVersion":"vm.neon.tech/v1","resourceVersion":"100683141"}, "reason": "Failed"}
</pre>
</details>

So I think there's maybe two issues here:

1. Something about this `"a used vhost backend has no free memory slots left"` — the "root cause"
2. `neonvm-controller` is writing a response with 200, `{ ... "allowed": true }` before it actually does the operation — causing the autoscaler-agent to think the operation was successful, when it hadn't actually started.

(2) can be worked around if need be, although it'd be nice for NeonVM to wait to respond until after it's tried the operation. (1) seems like an issue with NeonVM.

**Comment by cicdteam** [comment]
@sharnoff 

>Something about this "a used vhost backend has no free memory slots left" — the "root cause"

yes, seems so. If VM was created with `slotSize: 1Gi` and we are trying add 60+ DIIMMs into VM we got that error. Need to check how many memory slots able to create.

As workaround - setup `slotSize: 4Gi`, then for 16 CPU 64Gb RAM scale factor will be 16 (16 memory slots in use)

**Comment by vadim2404** [comment]
closed in favor of #350
